File: examples/04-kafka/python/dlq_after_retry_exhaustion.py
from datetime import timedelta

from temporalio import workflow
from temporalio.common import RetryPolicy
from temporalio.exceptions import ActivityError

# `validate` and `publish_to_dlq` are module-level @activity.defn functions.


@workflow.defn
class DlqRoutingWorkflow:
    # Temporal owns the retry series. Once `validate` exhausts its maximum_attempts,
    # the await raises an ActivityError here — that's our signal to route the
    # poison message to a Kafka DLQ instead of crashing the Workflow.
    @workflow.run
    async def process(self, order_id: str) -> None:
        try:
            await workflow.execute_activity(
                validate,
                order_id,
                start_to_close_timeout=timedelta(seconds=30),
                retry_policy=RetryPolicy(maximum_attempts=5),
            )
        except ActivityError as exhausted:
            await workflow.execute_activity(
                publish_to_dlq,
                args=[order_id, str(exhausted)],
                start_to_close_timeout=timedelta(seconds=10),
            )

File: examples/04-kafka/go/dlq_after_retry_exhaustion.go
package kafka

import (
	"time"

	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

// Temporal owns the retry series. Once Validate exhausts its MaximumAttempts the
// Future's Get returns an error here — that's our signal to route the poison
// message to a Kafka DLQ instead of failing the Workflow.
func DlqRoutingWorkflow(ctx workflow.Context, orderID string) error {
	orderCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Second,
		RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: 5},
	})
	if err := workflow.ExecuteActivity(orderCtx, "Validate", orderID).Get(orderCtx, nil); err != nil {
		dlqCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
			StartToCloseTimeout: 10 * time.Second,
		})
		return workflow.ExecuteActivity(dlqCtx, "PublishToDLQ", orderID, err.Error()).Get(dlqCtx, nil)
	}
	return nil
}
